torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -88,34 +92,62 @@ void ProcessGroupAgent::join() {
 }
 
-void ProcessGroupAgent::sync() {
-  // Block until all processes wants to sync. This is necessary before acquiring
-  // the lock below, because other processes might not enter sync() until it
-  // gets some response from this RpcAgent.
-  pg_->barrier()->wait();
-  // Acquire the lock on the send queue to prevent additional messages to be put
-  // onto the send queue.
-  std::unique_lock<std::mutex> lock(sendQueueMutex_);
-  // Wait until the send queue is depleted.
-  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
-  // Use another barrier in case different RpcAgent handles different amounts of
-  // workloads.
+void ProcessGroupAgent::waitAll() {
+  // Intuitively, waitAll is like all workers in the same communication world
+  // reaching a consensus that they all want to stop talking to others,
+  // so that, on exiting waitAll, the communication world will be
+  // like brand new without any pending Rpc work.
+
+  // A message can be in one of these states,
+  //   1. A worker calls send, putting a req message queue A.
+  //   2. The req message is popped and sent to wire.
+  //   3. The req message is received from wire and put into queue B.
+  //   4. The req message is popped and processed, a rep message is put
+  //      into queue B.
+  //   6. The rep message is popped and sent to wire.
+  //   7. The rep message is received from wire and put into queue A.
+  //   8. The rep is popped, marking the request future as complete.
+
+  // The assumptions of this solution are
+  //   1. Only one thread calls send, waitAll, join.
+  //   2. The state of a message only transfers in one-way.
+  //   3. Every worker has only limited work, so there is always a time that
+  //      on worker add a message in state 1.
+
+  // waitAll achieves this goal by waiting for the state space reducing,
+  // {12345678}, to {2345678}, to {345678}, to {45678}, ..., and eventually
+  // to an empty state space, {}.
+
+  // Assuming the state of a message can keep tranfering in best effort, this
+  // is to wait for the state space to collapse to an empty state space, {}.
+  // Note that the lifetime of a request Future span from state 1 to state 8,
+  // so exsitence of any Future in futures_ map is equivalent to exsitence
+  // of any message in the communication world, except for the TERMINATION
+  // message, which is not inserted into the futures_ map and does not
+  // ask for ACK or response.
+  waitSelf();
+
+  // This is to ensure no worker will add send work to their own thread pool
+  // queue before the shared goal is achived, elliminating the possibility of
+  // adding a message in state 1. Otherwise, a quick peer could exit waitAll
+  // and send again, disrruping the assumption when others are calling wait.
   pg_->barrier()->wait();
 }
 
+void ProcessGroupAgent::waitSelf() {
+    std::unique_lock<std::mutex> lock{futureMutex_};
+    futureDecreaseCV_.wait(lock, [&](){return futures_.empty();});
```

Comment:
Do we need to wait for the `ThreadPool` to complete as well? What if all of A's futures are cleared (meaning all A' requests are resolved), but there are still some `cb_` running which might send a request later. Will this be a problem?

Reply:
@mrshenli 

You are right.

I didn't make the purpose of this API clear.
They should be named as
`waitForSelfRequests()`
`waitForAllRequests()`

I will add code comments to clearly state that it only waits for all requests form itself are resolved. In the meanwhile, it could still be sending out RPC responses.

This is useful when during training, some following actions on a worker must depends on previous commands for other workers have been accepted and no exception happened. Then it could move on to the next set of commands. 
In the meanwhile, it could serve(respond with) values, like RRef, as other workers want, because **it doesn't care other's progress (not requiring a group sync, just need to meet local code execution dependency)**. 
For example, during an iteration, OR during an epoch before parameter synchronization happens.

I believe there will be cases that a worker also want all other's requests are resolved, for example, before checkpointing (or after an epoch or after a several iterations), the whole group should be in an intact and consistent state.

So for different levels of synchronization, the full API set might be

`waitForSelfRequests()`   
`waitForAllRequests()`   <------ In this PR I only care about these 2

`waitForSelfResponses()`
`waitForAllResponses()`

`waitForSelfRequestsAndResponses()`
`waitForAllRequestsAndResponses()`

It looks to me `waitForAllRequests()` derives to `waitForAllResponses()`, and then derives to `waitForAllRequestsAndResponses()`.

So we only need,

`waitForSelfRequests()`   
`waitForSelfResponses()`
`waitForSelfRequestsAndResponses()`

`waitForAllRequestsAndResponses()`

Reply:
@pritamdamania87 @satgera

Let's continue the sync/join API discussion here. Do you see if there is an easy way for users to do this type of sync completely in application code, without using our sync/join API? It seems to me one process needs to maintain some global var or context and asks others to call in to update those var/context.

Reply:
@xush6528 Can't this synchronization be done using a master node? For example, in case of checkpointing, the master node can request each trainer to run for X iterations and once that is done, the master node can schedule a checkpoint? 

> This is useful when during training, some following actions on a worker must depends on previous commands for other workers have been accepted and no exception happened. Then it could move on to the next set of commands.

Could you give an example of something like this? 

Also, I don't think the synchronization/join() API belongs in RPCAgent. If we really want to have some sort of barrier across all workers, we can add an API to RPCAgent which waits for all previously submitted RPCs on this worker to complete. The user can call this API followed by ProcessGroupGloo.barrier() to synchronize across all workers.

Reply:
Few more thoughts on this. I feel RpcAgent should only deal with point to point communication and not collective communication/synchronization. If we really need such a synchronization, we can probably give a separate API to users (ex: `dist.sync_workers()`) which does the necessary operations underneath.

Reply:
I agree with Pritam, I too believe RpcAgent should only deal with point to point communication and hence join and sync don't belong here. To support join/sync like higher level collectives something like a pause and resume request to an endpoint can be added instead.